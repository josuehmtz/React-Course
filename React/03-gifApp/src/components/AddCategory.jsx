import { useState } from "react"

export const AddCategory = ({onNewCategory}) => {

    const [inputValue, setInputValue] = useState("")

    const onChangeHandler = ({target}) => {
        setInputValue(target.value);
    }

    const onSubmit = (event) => {
        event.preventDefault();
        const cleanValue = inputValue.trim();
        if (cleanValue.length <= 1) return;
        onNewCategory(cleanValue);
        // setCategories( categories => [inputValue, ...categories]);
        setInputValue('');
    }

    return (
        <form onSubmit={ onSubmit }>
            <input 
                type="text" 
                placeholder="Search GIFs..."
                value={inputValue}
                className="m-3 p-3 bg-white shadow-md h-10 w-50 border-2 border-blue-900 rounded-xl inline-flex" 
                onChange={ onChangeHandler }
            />
            <button type="submit" className="p-1 m-3 shadow-md h-10 w-16 bg-orange-300 font-semibold rounded-xl">Add</button>
        </form>
    )
}
