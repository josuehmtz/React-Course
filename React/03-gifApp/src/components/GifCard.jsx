import PropTypes from 'prop-types';

export const GifCard = ({title, url}) => {
  return (
    <div>
        <div className="p-4 bg-white rounded-3xl">
            <img src={url} alt={title}/>
            <p className="bg-white text-2md pt-2 text-center">{title}</p>
        </div>
    </div>
  )
}

GifCard.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

