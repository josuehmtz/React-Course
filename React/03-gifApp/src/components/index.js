export * from './AddCategory';
export * from './GifGrid';
export * from './GifCard';
export * from './GifApp';