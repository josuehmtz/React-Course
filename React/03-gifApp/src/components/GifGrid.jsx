import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifCard } from "./GifCard";
import { PropTypes } from "prop-types";

export const GifGrid = ({category}) => {

    const { images, isLoading } = useFetchGifs(category);
    
    return (
        <>
            <h3 className="tracking-widest p-3 text-3xl text-bold">{category}</h3>
            {
                isLoading && ( <h2 className="text-6xl">Loading...</h2> )
            }
            <div className="flex flex-wrap gap-6 flex-auto">
                {
                    images.map((image) => (
                        <GifCard key={image.id} {...image}/>
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

