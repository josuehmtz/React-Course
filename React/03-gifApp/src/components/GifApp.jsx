import { useState } from "react";
import { AddCategory  } from "./AddCategory";
import { GifGrid } from "./GifGrid";

export const GifApp = () => {

  const [categories, setCategories] = useState(['Venom']);

  const onAddCategory = (category) => {

    if (categories.includes(category)) return;
    setCategories([category, ...categories]);
    // setCategories( c => [...categories, 'Gravity Falls']);
  }

  return (
    <>
        <h1 className="text-blue-900 m-3 font-bold">GifsApp</h1>
        <div className="inline-flex">
          <AddCategory onNewCategory={onAddCategory}/>
        </div>
        {
          categories.map(category => (
            <GifGrid key={category} category={category}/>
          ))
        }

    </>
  )
}
