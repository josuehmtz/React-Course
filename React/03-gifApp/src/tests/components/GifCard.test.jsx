import { getByRole, getByText, render, screen } from "@testing-library/react";
import { GifCard } from "../../components";


describe('Pruebas en GifCard Component', () => {
    
    const url = 'http://localhost/';
    const title = 'Mexico';

   test('El titulo y la url deben son obligatorios', () => { 
        const {container, getAllByText} = render( <GifCard title={title} url={url}/> )
        // expect(container).toMatchSnapshot();
        expect(screen.getAllByText(title)).toBeTruthy();
        // expect(screen.getAllByText(url)).toBeTruthy();
    }) 

    test('Debe de mostrar el url y el alt adecuados', () => { 
        const {container} = render( <GifCard title={title} url={url}/> )
        // expect( screen.getByRole('img').src).toBe(url)
        // expect( screen.getByRole('img').alt).toBe(title)
        const {src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title) 
     })

     test('debe de mostra el titulo en el componente', () => { 
        render( <GifCard url={url} title={title}/> )
        expect(screen.getByText(title)).toBeTruthy();
      })
   
    test('Debe de hacer match con el snapshot', () => { 
        const {container} = render( <GifCard title={title} url={url}/> )
        expect(container).toMatchSnapshot();
    }) 
}); 