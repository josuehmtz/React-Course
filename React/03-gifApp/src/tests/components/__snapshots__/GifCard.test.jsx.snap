// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifCard Component Debe de hacer match con el snapshot 1`] = `
<div>
  <div>
    <div
      class="p-4 bg-white rounded-3xl"
    >
      <img
        alt="Mexico"
        src="http://localhost/"
      />
      <p
        class="bg-white text-2md pt-2 text-center"
      >
        Mexico
      </p>
    </div>
  </div>
</div>
`;
